d2layouts/d2near/layout.go

```diff
@@ -10,79 +10,139 @@ import (
 	"oss.terrastruct.com/d2/d2graph"
 	"oss.terrastruct.com/d2/lib/geo"
 	"oss.terrastruct.com/d2/lib/label"
-	"oss.terrastruct.com/util-go/go2"
 )
 
 const pad = 20
 
 // Layout finds the shapes which are assigned constant near keywords and places them.
-func Layout(ctx context.Context, g *d2graph.Graph, constantNears []*d2graph.Object) error {
-	if len(constantNears) == 0 {
+func Layout(ctx context.Context, g *d2graph.Graph, constantNearGraphs []*d2graph.Graph) error {
+	if len(constantNearGraphs) == 0 {
 		return nil
 	}
 
+	for _, tempGraph := range constantNearGraphs {
+		tempGraph.Root.ChildrenArray[0].Parent = g.Root
+		for _, obj := range tempGraph.Objects {
+			obj.Graph = g
+		}
+	}
+
 	// Imagine the graph has two long texts, one at top center and one at top left.
 	// Top left should go left enough to not collide with center.
 	// So place the center ones first, then the later ones will consider them for bounding box
 	for _, processCenters := range []bool{true, false} {
-		for _, obj := range constantNears {
+		for _, tempGraph := range constantNearGraphs {
+			obj := tempGraph.Root.ChildrenArray[0]
 			if processCenters == strings.Contains(d2graph.Key(obj.Attributes.NearKey)[0], "-center") {
+				prevX, prevY := obj.TopLeft.X, obj.TopLeft.Y
 				obj.TopLeft = geo.NewPoint(place(obj))
+				dx, dy := obj.TopLeft.X-prevX, obj.TopLeft.Y-prevY
+
+				for _, subObject := range tempGraph.Objects {
+					// `obj` already been replaced above by `place(obj)`
+					if subObject == obj {
+						continue
+					}
+					subObject.TopLeft.X += dx
+					subObject.TopLeft.Y += dy
+				}
+				for _, subEdge := range tempGraph.Edges {
+					for _, point := range subEdge.Route {
+						point.X += dx
+						point.Y += dy
+					}
+				}
+
+				g.Edges = append(g.Edges, tempGraph.Edges...)
```

Comment:
why not add objects back the same way? why do the `attachChildren` way?

Reply:
my bad, deleted `attachChildren` should be ok, `tempGraph` contain all the objects, no need to do the recursive way, and also i place `g.Edges = append(g.Edges, tempGraph.Edges...)` to second loop, which is now for attach objects and edges to original graph